akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -139,34 +142,45 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
 
         if (isDebugEnabled && newState.name != oldState.name) recordStateChange(oldState.name, newState.name)
         if (allDataFlushed(newState)) onAllDataFlushed()
+        allowReadingIncomingFrames(controlFrameBuffer.size < settings.outgoingControlFrameBufferSize)
       }
 
-      private[http2] sealed trait MultiplexerState extends Product {
+      sealed trait MultiplexerState extends Product {
         def name: String = productPrefix
 
         def onPull(): MultiplexerState
+        @silent("references private")
         def pushControlFrame(frame: FrameEvent): MultiplexerState
         def connectionWindowAvailable(): MultiplexerState
         def enqueueOutStream(streamId: Int): MultiplexerState
         def closeStream(streamId: Int): MultiplexerState
 
-        protected def sendDataFrame(streamId: Int, sendableOutstreams: immutable.Set[Int]): MultiplexerState = {
+        protected def sendDataFrame(streamId: Int): MultiplexerState = {
           val maxBytesToSend = currentMaxFrameSize min connectionWindowLeft
           val result = pullNextFrame(streamId, maxBytesToSend)
-          val frame = result.frame
-          pushFrameOut(frame)
-          connectionWindowLeft -= frame.payload.length
+          def send(frame: DataFrame): Unit = {
+            pushFrameOut(frame)
+            connectionWindowLeft -= frame.payload.length
+          }
 
           result match {
-            case PullFrameResult.SendFrame(_, hasMore) =>
-              if (hasMore) WaitingForNetworkToSendData(sendableOutstreams + streamId)
-              else {
-                val remainingStreams = sendableOutstreams - streamId
-                if (remainingStreams.isEmpty) Idle
-                else WaitingForNetworkToSendData(remainingStreams)
+            case PullFrameResult.SendFrame(frame, hasMore) =>
+              send(frame)
+              if (hasMore) {
+                sendableOutstreams += streamId
+                WaitingForNetworkToSendData
+              } else {
+                if (sendableOutstreams.isEmpty) Idle
+                else WaitingForNetworkToSendData
               }
-            case PullFrameResult.SendFrameAndTrailer(_, trailer) =>
-              WaitingForNetworkToSendControlFrames(Vector(trailer), sendableOutstreams - streamId)
+            case PullFrameResult.SendFrameAndTrailer(frame, trailer) =>
+              send(frame)
+              controlFrameBuffer += trailer
+              WaitingForNetworkToSendControlFrames
+            case PullFrameResult.NothingToSend =>
```

Comment:
We try to avoid removing elements from the queue to avoid scanning in `closeStream` below, so we just leave it in there and added a new return value of `NothingToSend` for `pullNextFrame` to be able to skip those entries here.

Reply:
So the motivation is that looking for the stream in `closeStream` is more expensive than just keeping it in there until we encounter it here and go to `WaitingForData` / `WaitingForNetworkToSendData.onPull()`. That's not super obvious to me but I believe you :).

Reply:
> So the motivation is that looking for the stream in `closeStream` is more expensive than just keeping it in there until we encounter it here and go to `WaitingForData` / `WaitingForNetworkToSendData.onPull()`

Yes.

I guess it depends on how often `closeStream` is called. The intuition would say, probably not so often, as it's only needed if a stream is cancelled and that should be exceptional. So, removing single elements in `O(n)` would be fine as long as not a whole bunch of streams out of many is removed and we run into a `O(n^2)` situation.

Maybe I'll run another benchmark and see if that would happen in normal situations.

Reply:
I found another reason for the current logic:

With the queues we allow duplicates in the queue to be able to keep `enqueueOutStream` idempotent (without having to scan the queue for existing elements). We currently might add duplicates to the queue e.g. in `OutStream.increaseWindow`.

To get rid of `NothingToSend`, we would have to ensure that a stream is never put into the queue twice by keeping a flag on the `OutStream` side whether we are currently enqueued or not. Not sure if that's worth it right now.

Reply:
I added #3882 to make an attempt to revisit this complexity.